server/src/Main.hs

```diff
@@ -189,7 +189,7 @@ completeLogin origin rpIdHash db pending = do
   Scotty.liftAndCatchIO $ TIO.putStrLn $ "Login complete <= " <> jsonText (WA.stripRawCredential cred)
 
   options <-
-    Scotty.liftAndCatchIO (getPendingOptions pending cred) >>= \case
+    Scotty.liftAndCatchIO (getPendingOptions (pendingLogins pending) (WA.araClientData $ WA.cResponse cred)) >>= \case
```

Comment:
These calls are getting a bit long. I think it would be better for the `PendingOps` module to provide the same calls as before, but specialized to the two ceremony kinds:
```haskell
insertPendingRegistration ::
  PendingOps ->
  (WA.Challenge -> WA.CredentialOptions 'Registration) ->
  IO (WA.CredentialOptions 'Registration)

getPendingRegistration ::
  PendingOps ->
  WA.Credential 'Registration raw ->
  IO (Either String (WA.CredentialOptions 'Registration))
```

And the same with `s/Registration/Authentication`

Reply:
Yeah, those signatures are more straightforward. The singletons and polymorphism made the logic harder to follow. Multi-threading and cleaning up are still a problem.

What do you think of just dropping this module, and storing the challenges in the database?

Reply:
How is multi-threading and cleanup a problem? Multi-threading shouldn't be a problem with the use of STM, and cleanup is done regularly too by this module.

Regarding storing these operations in the database, we talked about this before. You mentioned that storing it in memory wouldn't work for you because there's multiple web servers handling requests. However, note that both requests (for beginning and ending of a register/login) happen in the same javascript code, e.g. here: https://github.com/tweag/webauthn/blob/084a264d84e6629e3b1b9bc068840345b82ba7eb/server/www/unauthenticated.js#L15-L40

I believe [according to MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Session) these requests always go to the same server:

> In client-server protocols, like HTTP, sessions consist of three phases:
> 
> 1. The client establishes a TCP connection (or the appropriate connection if the transport layer is not TCP).
> 2. The client sends its request, and waits for the answer.
> 3. The server processes the request, sending back its answer, providing a status code and appropriate data.
> 
> As of HTTP/1.1, the connection is no longer closed after completing the third phase, and the client is now granted a further request: this means the second and third phases can now be performed any number of times.

This seemingly indicates that it's always going to be the same web server receiving the response, so I don't think we have to worry about that.

But here are some pros and cons I can think of for storing it in a database:
- [Pro] Lookup efficiency: Database engines have lots of optimizations to make lookups fast
- [Con] Disk access delay: On the other hand, databases need to read from and write to the disk, which comes with a big delay compared to memory
- [Pro] Durability: If a user starts a login, but then the server crashes, the user can still finish their login
- [Con] Complexity: It requires updating the database schema and makes updates to it harder
- [Pro] Less memory usage: Storing something on the database instead means you use less memory

Personally I don't value durability very high in this case, as it's mainly just a randomly generated challenge to prevent replay attacks, which can just be regenerated. This leaves me with the impression that neither approach is fundamentally better, though for the sake of this demo I think we should value simplicity more.

I do think however that the library should be more clear about what exactly needs to be stored for later, so that for both in-memory storage and database storage there's no unneeded information stored.

Reply:
> This leaves me with the impression that neither approach is fundamentally better, though for the sake of this demo I think we should value simplicity more.

I agree that we should value simplicity more. The singletons and multi-threading primitives require a reader to know the concepts beyond the basic web development knowledge. The breakdown of pros and cons in your comment comes from the perspective of the run-time properties. However, the server has two ways to store data: database and in-memory, and that's complicated.

Also, I'd value durability higher, since without it there would hiccups at the register/login flow. The assumption of the same server doing a response gets more fragile with complex server-side network structure. It is up to the load balancer how the requests get distributed between the application servers, even if it is the same balancer server receiving the request. Besides, the pending operations would get lost on a deploy. So, storing challenges in a database would help both with durability and simplicity.

I'm less concerned about durability of the example server, but I wish it were easier to understand and tell the details of implementation apart from the core logic of working with webauthn.